web/auth/client.go

```
@@ -0,0 +1,154 @@
+type ClientRegistrationError struct {
+	Code        int    `json:"-"`
+	Error       string `json:"error"`
+	Description string `json:"error_description,omitempty"`
+}
+
+// Create is a function that sets some fields, and then save it in Couch.
+func (c *Client) Create(i *instance.Instance) *ClientRegistrationError {
+	if len(c.RedirectURIs) == 0 {
+		return &ClientRegistrationError{
+			Code:        http.StatusBadRequest,
+			Error:       "invalid_redirect_uri",
+			Description: "redirect_uris is mandatory",
+		}
+	}
+	for _, redirectURI := range c.RedirectURIs {
+		u, err := url.Parse(redirectURI)
+		if err != nil ||
+			(u.Scheme != "http" && u.Scheme != "https") ||
+			u.Host == i.Domain ||
+			u.Fragment != "" {
+			return &ClientRegistrationError{
+				Code:        http.StatusBadRequest,
+				Error:       "invalid_redirect_uri",
+				Description: fmt.Sprintf("%s is invalid", redirectURI),
+			}
+		}
+	}
+	if c.ClientName == "" {
+		return &ClientRegistrationError{
+			Code:        http.StatusBadRequest,
+			Error:       "invalid_client_metadata",
+			Description: "client_name is mandatory",
+		}
+	}
+	if c.SoftwareID == "" {
+		return &ClientRegistrationError{
+			Code:        http.StatusBadRequest,
+			Error:       "invalid_client_metadata",
+			Description: "software_id is mandatory",
+		}
+	}
+
+	c.CouchID = ""
+	c.CouchRev = ""
+	c.ClientID = ""
+	c.ClientSecret = ""
+	c.SecretExpiresAt = 0
+	c.RegistrationToken = ""
+	c.GrantTypes = []string{"authorization_code", "refresh_token"}
+	c.ResponseTypes = []string{"code"}
+
+	if err := couchdb.CreateDoc(i, c); err != nil {
+		return &ClientRegistrationError{
+			Code:  http.StatusInternalServerError,
+			Error: "internal_server_error",
+		}
+	}
+
+	var err error
+	c.ClientSecret, err = crypto.NewJWT(i.HmacSecret, jwt.StandardClaims{
+		Audience: "client_secret",
+		Issuer:   i.Domain,
+		IssuedAt: time.Now().Unix(),
+		Subject:  c.CouchID,
+	})
+	if err != nil {
+		return &ClientRegistrationError{
```

Error swallowing : we should at least log the original error